src/dune_patch/dune_patch.ml

```diff
@@ -0,0 +1,90 @@
+open Stdune
+module Re = Dune_re
+
+include struct
+  open Dune_engine
+  module Action = Action
+  module Process = Process
+  module Utils = Utils
+end
+
+include struct
+  open Dune_lang
+  module Value = Value
+end
+
+let re =
+  Re.compile
+    (Re.non_greedy (Re.seq [ Re.str {|--- a/|}; Re.group (Re.rep1 Re.any); Re.eol ]))
+;;
+
+let files_of_patch ~dir patch =
+  Re.all re patch
+  |> List.map ~f:(fun group ->
+    let group = Re.Group.get group 1 in
+    let local_path = Path.Local.of_string group in
+    Path.append_local dir local_path)
+;;
+
+let files_of_patch_file ~dir path = Io.read_file path |> files_of_patch ~dir
+
+module Spec = struct
+  type ('path, 'target) t = 'path
+
+  let name = "patch"
+  let version = 1
+  let bimap p f _ = f p
+  let is_useful_to ~distribute:_ ~memoize = memoize
+
+  let encode (p : (_, _) t) input _ : Dune_lang.t =
+    List [ Dune_lang.atom_or_quoted_string name; input p ]
+  ;;
+
+  let action p ~(ectx : Action.Ext.context) ~(eenv : Action.Ext.env) =
+    let open Fiber.O in
+    let* () = Fiber.return () in
+    let input = Value.to_string ~dir:eenv.working_dir (Value.Path p) in
+    let patch =
+      let path = eenv.env |> Env_path.path in
+      let program = "patch" in
+      match Bin.which ~path program with
+      | Some p -> p
+      | None ->
+        Utils.program_not_found
+          ~loc:None
+          ?context:(Option.map ~f:(fun ctx -> ctx.name) ectx.context)
+          program
+    in
+    (* 1. Work out which files are being patched. *)
+    let files = files_of_patch_file ~dir:eenv.working_dir p in
+    (* 2. Materialize those files. *)
+    List.iter files ~f:(fun file ->
+      Io.copy_file ~src:file ~dst:(Path.extend_basename file ~suffix:".for_patch") ());
+    List.iter files ~f:(fun file ->
+      Unix.rename (Path.to_string file ^ ".for_patch") (Path.to_string file));
+    (* 3. Apply the patch. *)
+    (* TODO opam has a preprocessing step that we should probably apply *)
+    Process.run
+      ~dir:eenv.working_dir
+      ~display:Dune_engine.Display.Quiet
```

Comment:
I think this should be inherited from `Clflags.display` - not in this module, but in whoever is calling this function.

Reply:
Would it make sense to add display to eenv?

Reply:
Might be. Let's not worry about this now.